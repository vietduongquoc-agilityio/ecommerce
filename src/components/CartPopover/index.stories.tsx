import { Meta, StoryObj } from "@storybook/react";
import CartPopover from "./";

const meta: Meta<typeof CartPopover> = {
  title: "Components/CartPopover",
  component: CartPopover,
  tags: ["autodocs"],
  parameters: {
    layout: "fullscreen",
  },
};

export default meta;

type Story = StoryObj<typeof CartPopover>;

export const Default: Story = {};
