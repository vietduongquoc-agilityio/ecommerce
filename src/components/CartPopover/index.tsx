"use client";

import * as Dialog from "@radix-ui/react-dialog";
import { useRouter } from "next/navigation";
import Image from "next/image";
import styles from "./styles.module.css";
import CartIcon from "@/assets/Images/CartIcon.png";
// import Product1 from "@/assets/Images/Product1.png";
// import Product2 from "@/assets/Images/Product2.png";
import { colors, fonts } from "@/themes";
import DeleteVector from "@/assets/Images/DeleteVector.png";
import Button from "../Button";
import { useCartStore } from "@/stores/cartStore";

const CartPopover = () => {
  const router = useRouter();
  const { cart } = useCartStore();

  const calculateSubtotal = () => {
    return cart.reduce((total, item) => total + item.price * item.quantity, 0);
  };

  return (
    <Dialog.Root>
      <Dialog.Trigger asChild>
        <Button variant="ghost">ðŸ›’ Cart ({cart.length})</Button>
      </Dialog.Trigger>

      <Dialog.Portal>
        {/* Overlay */}
        <Dialog.Overlay className={styles.overlay} />

        {/* Cart Popover Content */}
        <Dialog.Content className={styles.content}>
          <div className={styles.header}>
            <Dialog.Title className={styles.title}>Shopping cart</Dialog.Title>

            {/* Close Button */}
            <Dialog.Close asChild>
              <Image
                src={CartIcon}
                alt="CartIcon"
                className={styles.closeButton}
              />
            </Dialog.Close>
          </div>

          <div style={{ padding: "40px" }}>
            {/* Cart Items */}
            {cart.length > 0 ? (
              cart.map((item) => (
                <div
                  key={item.documentId}
                  style={{
                    display: "flex",
                    alignItems: "center",
                    marginBottom: "30px",
                    justifyContent: "space-between",
                  }}
                >
                  <Image
                    src={item.image}
                    alt={item.name}
                    width={105}
                    height={105}
                    style={{ borderRadius: fonts.borderRadius.sm }}
                  />
                  <div style={{ width: "130px" }}>
                    <p
                      style={{
                        fontWeight: fonts.weight.regular,
                        fontSize: fonts.size.xs,
                        color: colors.black,
                      }}
                    >
                      {item.name}
                    </p>
                    <div
                      style={{
                        display: "flex",
                        justifyContent: "space-between",
                        marginTop: "15px",
                        alignItems: "center",
                      }}
                    >
                      <p style={{ fontWeight: "300", fontSize: fonts.size.xs }}>
                        {item.quantity}
                      </p>
                      <p style={{ fontWeight: "300", fontSize: "12px" }}>X</p>
                      <p
                        style={{
                          fontWeight: fonts.weight.medium,
                          fontSize: "12px",
                          color: colors.primary.yellow,
                        }}
                      >
                        Rs. {item.price.toLocaleString("id-ID")}
                      </p>
                    </div>
                  </div>
                  <Image
                    className={styles.cartIcon}
                    src={DeleteVector}
                    alt="DeleteVector"
                  />
                </div>
              ))
            ) : (
              <p style={{ textAlign: "center", marginTop: "20px" }}>
                Your cart is empty.
              </p>
            )}

            {/* Subtotal */}
            <div
              style={{
                display: "flex",
                justifyContent: "space-between",
                width: "300px",
                marginTop: "20px",
              }}
            >
              <p
                style={{
                  fontWeight: fonts.weight.regular,
                  fontSize: fonts.size.xs,
                  color: colors.black,
                }}
              >
                Subtotal
              </p>
              <p
                style={{
                  fontWeight: fonts.weight.semiBold,
                  fontSize: fonts.size.xs,
                  color: colors.primary.yellow,
                }}
              >
                Rs. {calculateSubtotal().toLocaleString("id-ID")}
              </p>
            </div>
          </div>

          {/* Action Buttons */}
          <div
            style={{
              width: "100%",
              display: "flex",
              padding: "32px 40px",
              borderTop: "solid 1px #d9d9d9",
            }}
          >
            <Button
              variant="outline"
              style={{
                padding: "6px 30px",
                borderRadius: "50px",
                fontWeight: fonts.weight.regular,
                fontSize: "12px",
                marginRight: "20px",
              }}
              onClick={() => router.push("/cart")}
            >
              Cart
            </Button>

            <Button
              variant="outline"
              style={{
                padding: "6px 30px",
                borderRadius: "50px",
                fontWeight: fonts.weight.regular,
                fontSize: "12px",
              }}
              onClick={() => router.push("/checkout")}
            >
              Checkout
            </Button>
          </div>
        </Dialog.Content>
      </Dialog.Portal>
    </Dialog.Root>
  );
};

export default CartPopover;
