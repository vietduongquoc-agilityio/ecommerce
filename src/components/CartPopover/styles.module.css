.cartIcon {
  cursor: pointer;
  transition: transform 0.3s ease;
}

.cartIcon:hover {
  transform: scale(1.1);
}

.overlay {
  background-color: rgba(0, 0, 0, 0.5);
  position: fixed;
  inset: 0;
  animation: fadeIn 150ms ease-in-out;
}

.labelInput {
  font-size: 16px;
  font-weight: 500;
  color: black;
  margin-bottom: 23px;
  margin-top: 22px;
}

.contentInput {
  font-size: 16px;
  font-weight: 500;
  /* height: 78px; */
  border-radius: 10px;
  background: white;
  border: solid 1px #9f9f9f;
  padding: 25px 10px;
}

.content {
  position: fixed;
  right: 0;
  top: 0;
  width: 420px;
  background: white;
  box-shadow: var(--shadow-6);
  transition: transform 0.3s ease-in-out;
  --font-family: var(--font-poppins);
  font-family: var(--font-family);
}

.header {
  padding: 30px 0;
  margin-left: 30px;
  --font-family: var(--font-poppins);
  border-bottom: solid 1px #d9d9d9;
  width: 290px;
}

.title {
  font-size: 24px;
  font-weight: 600;
  color: var(--mauve-12);
}

.description {
  font-size: 15px;
  color: var(--mauve-11);
  --font-family: var(--font-poppins);
}

.closeButton {
  position: absolute;
  top: 32px;
  right: 40px;
  background: none;
  border: none;
  font-size: 20px;
  cursor: pointer;
  color: var(--gray-6);
}

.closeButton:hover {
  color: var(--gray-9);
}

@keyframes fadeIn {
  from {
    opacity: 0;
  }
  to {
    opacity: 1;
  }
}
