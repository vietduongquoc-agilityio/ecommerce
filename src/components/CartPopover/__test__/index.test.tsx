import { render } from "@testing-library/react";
import CartPopover from "../index";
import { useRouter } from "next/navigation";

jest.mock("next/navigation", () => ({
  useRouter: jest.fn(),
}));

const mockPush = jest.fn();
describe("CartPopover Component", () => {
  beforeEach(() => {
    jest.clearAllMocks();

    (useRouter as jest.Mock).mockReturnValue({ push: mockPush });
  });
  it("renders correctly and matches snapshot", () => {
    const { asFragment } = render(<CartPopover />);
    expect(asFragment()).toMatchSnapshot();
  });
});
