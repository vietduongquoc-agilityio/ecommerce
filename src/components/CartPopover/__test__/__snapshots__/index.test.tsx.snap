// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CartPopover Component renders correctly and matches snapshot 1`] = `
<DocumentFragment>
  <img
    alt="cart-icon"
    aria-controls="radix-:r0:"
    aria-expanded="false"
    aria-haspopup="dialog"
    class="cartIcon"
    data-nimg="1"
    data-state="closed"
    decoding="async"
    height="22"
    loading="lazy"
    src="/_next/image?url=%2Fimg.jpg&w=64&q=75"
    srcset="/_next/image?url=%2Fimg.jpg&w=32&q=75 1x, /_next/image?url=%2Fimg.jpg&w=64&q=75 2x"
    style="color: transparent;"
    type="button"
    width="25"
  />
</DocumentFragment>
`;
